src/main/java/net/snowflake/ingest/streaming/internal/DataValidationUtil.java

```diff
@@ -412,15 +419,34 @@ static String validateAndParseString(
           "STRING",
           new String[] {"String", "Number", "boolean", "char"});
     }
-    int maxLength = maxLengthOptional.orElse(BYTES_16_MB);
+    byte[] utf8Bytes = output.getBytes(StandardCharsets.UTF_8);
 
-    if (output.length() > maxLength) {
+    // Strings can never be larger than 16MB
+    if (utf8Bytes.length > BYTES_16_MB) {
       throw valueFormatNotAllowedException(
           columnName,
           input,
           "STRING",
-          String.format("String too long: length=%d maxLength=%d", output.length(), maxLength));
+          String.format(
+              "String too long: length=%d bytes maxLength=%d bytes",
+              utf8Bytes.length, BYTES_16_MB));
     }
+
+    // If max allowed length is specified (e.g. VARCHAR(10)), the number of unicode characters must
+    // not exceed this value
+    maxLengthOptional.ifPresent(
+        maxAllowedCharacters -> {
+          int actualCharacters = output.codePointCount(0, output.length());
```

Comment:
What does `codePointCount` do?

Reply:
"The codePointCount() method is used to count the number of Unicode code points in the specified text range of a given String.

The text range begins at the specified beginIndex and extends to the char at index endIndex - 1. Thus the length (in chars) of the text range is endIndex-beginIndex. Unpaired surrogates within the text range count as one code point each."

Reply:
Yeah, this calculates the number of unicode characters in the string, which we need for `VARCHAR(N)`. For example:

```java
String s = "🍞";
Assert.assertEquals(2, s.length());
Assert.assertEquals(1, s.codePointCount(0, s.length()));
```